### PR TITLE
Unified cache for s3, hdfs, url, hive table functions and table engines, remove old hive cache

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Unified cache for `s3`, `hdfs`, `url`, `hive` table functions and table engines, remove old hive cache

### CODE REVIEW

File: src/Interpreters/Cache/FileCache.cpp [hunk 1]
```diff
@@ -79,6 +77,37 @@ void FileCache::removeKeyDirectoryIfExists(const Key & key, std::lock_guard<std:
         fs::remove(key_prefix_path);
 }
 
+void FileCache::removeEmptyKey(const Key & key, std::lock_guard<std::mutex> & cache_lock, const CacheKeyRemoveSettings & remove_settings)
+{
+    auto it = files.find(key);
+    if (it == files.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "There is not key `{}` in cache", key.toString());
+
+    assert(files.at(key).file_segments_by_offset.empty());
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:86]:
```suggestion
    assert(it->file_segments_by_offset.empty());
```

File: src/Interpreters/Cache/FileCache.cpp [hunk 2]
```diff
@@ -79,6 +77,37 @@ void FileCache::removeKeyDirectoryIfExists(const Key & key, std::lock_guard<std:
...
+        if (on_key_eviction_func)
+        {
+            try
+            {
+                on_key_eviction_func();
+            }
+            catch (...)
+            {
+                tryLogCurrentException(__PRETTY_FUNCTION__);
+                assert(false);
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:100]:
```suggestion
                chassert(false);
```

File: src/Interpreters/Cache/FileCache.cpp [hunk 3]
```diff
@@ -418,6 +451,11 @@ FileSegmentsHolder FileCache::getOrSet(const Key & key, size_t offset, size_t si
     {
         fillHolesWithEmptyFileSegments(file_segments, key, range, /* fill_with_detached */false, settings, cache_lock);
     }
+
+    auto & on_key_eviction_func = files[key].on_key_eviction_func;
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:455]: Put this into `CreateFileSegmentSettings` instead and execute in `addCell` when a new key is created.

File: src/Interpreters/Cache/FileCache.cpp [hunk 4]
```diff
@@ -456,32 +494,48 @@ FileSegmentsHolder FileCache::get(const Key & key, size_t offset, size_t size)
...
 
-    if (files[key].contains(offset))
+    auto & offsets = files[key].file_segments_by_offset;
+
+    if (offsets.empty())
+    {
+        auto key_path = getPathInLocalCache(key);
+
+        if (!fs::exists(key_path))
+            fs::create_directories(key_path);
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.cpp:516]: Make a separate method `initializeNewKey()` and add `on_key_eviction_func` there?

File: src/Interpreters/Cache/FileCache.h [hunk 5]
```diff
@@ -24,6 +24,12 @@
 namespace DB
 {
 
+struct CacheKeyRemoveSettings
+{
+    bool call_key_removal_callback = true;
```
Comment (kssenii) [inline src/Interpreters/Cache/FileCache.h:29]: get rid of this

File: src/Storages/Hive/StorageHive.cpp [hunk 6]
```diff
@@ -201,42 +201,28 @@ class StorageHiveSource : public ISource, WithContext
...
 
-                String uri_with_path = hdfs_namenode_url + current_path;
-                auto compression = chooseCompressionMethod(current_path, compression_method);
-                std::unique_ptr<ReadBuffer> raw_read_buf;
+                String uri_with_path = fs::path(hdfs_namenode_url) / current_path;
+                std::unique_ptr<ReadBufferFromFileBase> remote_read_buf;
                 try
                 {
-                    auto get_raw_read_buf = [&]() -> std::unique_ptr<ReadBuffer>
+                    remote_read_buf = StorageHDFS::createHDFSReadBuffer(hdfs_namenode_url, current_path, nullptr, getContext());
```
Comment (kssenii) [inline src/Storages/Hive/StorageHive.cpp:216]: Revert these changes?

File: src/Storages/wrapWithCachedReadBuffer.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    bool modification_time_changed = getModificationTimeCache().updateModificationTimeIfChanged(cache_key, object_modification_time);
+    if (modification_time_changed)
+    {
+        LOG_TEST(
+            &Poco::Logger::get("wrapeWithCachedReadBuffer"),
+            "Removing `{}` from cache, because modification time changed",
+            cache_key.toString());
+
+        CacheKeyRemoveSettings settings{ .call_key_removal_callback = false };
+        /// FIXME: it will remove even if the file segments are currently used by someone.
```
Comment (kssenii) [inline src/Storages/wrapWithCachedReadBuffer.cpp:98]: TODO

File: src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp [hunk 8]
```diff
@@ -516,6 +519,11 @@ CachedOnDiskReadBufferFromFile::~CachedOnDiskReadBufferFromFile()
     {
         appendFilesystemCacheLog((*current_file_segment_it)->range(), read_type);
     }
+
+    /// If getOrSet() method was not called or failed with exception before we put key in cache,
+    /// we will not call on_key_eviction_func, which will result in memory leaks, so do it here.
+    if (on_key_eviction_func && !cache->hasKey(cache_key))
```
Comment (kssenii) [inline src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp:525]: Check if `hasKey` is dangerous to use because a race is possible

**Comment by tavplubix** [comment]
All changes related to hive are blocked by https://github.com/ClickHouse/ClickHouse/issues/43541 (the tests are temporarily disabled, so we cannot merge new features)
